src/probe/probe.c

```diff
@@ -15,42 +24,551 @@
+			probe_dma[i].stream_tag;
+		_probe->inject_dma[first_free].dma_buffer_size =
+			probe_dma[i].dma_buffer_size;
+
+		err = probe_dma_init(&_probe->inject_dma[first_free]);
+		if (err < 0) {
+			trace_probe_error("probe_dma_add() error: probe_dma_init() failed");
+			_probe->inject_dma[first_free].stream_tag =
+				PROBE_DMA_INVALID;
+			return err;
+		}
+	}
+
 	return 0;
 }
 
 int probe_dma_info(struct sof_ipc_probe_info_params *data, uint32_t max_size)
 {
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	tracev_probe("probe_dma_info()");
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_info() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	data->rhdr.hdr.size = sizeof(*data);
+
+	/* iterate over inject DMAs */
+	while (i < CONFIG_PROBE_DMA_MAX &&
+	       data->rhdr.hdr.size + sizeof(struct probe_dma) < max_size) {
+		/* save it if valid */
+		if (_probe->inject_dma[i].stream_tag != PROBE_DMA_INVALID) {
+			data->probe_dma[j].stream_tag =
+				_probe->inject_dma[i].stream_tag;
+			data->probe_dma[j].dma_buffer_size =
+				_probe->inject_dma[i].dma_buffer_size;
+			j++;
+			/* and increase reply header size */
+			data->rhdr.hdr.size += sizeof(struct probe_dma);
+		}
+
+		i++;
+	}
+
+	return 1;
+}
+
+static int is_probe_stream_used(uint32_t stream_tag)
+{
+	uint32_t i;
+
+	for (i = 0; i < CONFIG_PROBE_POINTS_MAX; i++) {
+		if (_probe->probe_points[i].stream_tag == stream_tag)
+			return 1;
+	}
+
 	return 0;
 }
 
 int probe_dma_remove(uint32_t count, uint32_t *stream_tag)
 {
+	uint32_t i;
+	uint32_t j;
+	int err;
+
+	tracev_probe("probe_dma_remove() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_dma_remove() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
+		tracev_probe("\tstream_tag[%u] = %u", i, stream_tag[i]);
+
+		if (is_probe_stream_used(stream_tag[i]))
+			return -EINVAL;
+
+		for (j = 0; j < CONFIG_PROBE_DMA_MAX; j++) {
+			if (_probe->inject_dma[j].stream_tag == stream_tag[i]) {
+				err = probe_dma_deinit(&_probe->inject_dma[j]);
+				if (err < 0)
+					return err;
+			}
+		}
+	}
+
 	return 0;
 }
 
 int probe_point_add(uint32_t count, struct probe_point *probe)
 {
+	uint32_t i;
+	uint32_t j;
+	uint32_t buffer_id;
+	uint32_t first_free;
+	uint32_t dma_found;
+	struct ipc_comp_dev *dev;
+
+	tracev_probe("probe_point_add() count = %u", count);
+
+	if (!_probe) {
+		trace_probe_error("probe_point_add() error: Not initialized.");
+
+		return -EINVAL;
+	}
+
+	for (i = 0; i < count; i++) {
+		tracev_probe("\tprobe[%u] buffer_id = %u, purpose = %u, stream_tag = %u",
+			     i, probe[i].buffer_id, probe[i].purpose,
+			     probe[i].stream_tag);
+
+		if (probe[i].purpose == PROBE_PURPOSE_EXTRACTION &&
+		    _probe->ext_dma.stream_tag == PROBE_DMA_INVALID) {
+			trace_probe_error("probe_point_add() error: Setting probe for extraction, while extraction DMA not enabled.");
+
+			return -EINVAL;
+		}
+
+		dev = ipc_get_comp_by_id(ipc_get(), probe[i].buffer_id);
```

Comment:
and here

Reply:
added but I think "get_comp_by_id" is self explanatory